ext/dom/token_list.c

```diff
@@ -0,0 +1,729 @@
+
+typedef struct dom_token_list_it {
+	zend_object_iterator it;
+	/* Store the hash position here to allow multiple (e.g. nested) iterations of the same token list. */
+	HashPosition pos;
+	php_libxml_cache_tag cache_tag;
+} dom_token_list_it;
+
+static zend_always_inline bool dom_contains_ascii_whitespace(const char *data)
+{
+	return strpbrk(data, ascii_whitespace) != NULL;
+}
+
+static zend_always_inline void dom_add_token(HashTable *ht, zend_string *token)
+{
+	/* Key outlives the value's lifetime because as long as the entry is in the table it is kept alive. */
+	zval zv;
+	ZVAL_STR(&zv, token);
+	zend_hash_add(ht, token, &zv);
+}
+
+/* https://dom.spec.whatwg.org/#concept-ordered-set-parser
+ * and https://infra.spec.whatwg.org/#split-on-ascii-whitespace */
+static void dom_ordered_set_parser(HashTable *token_set, const char *position)
+{
+	/* Adapted steps from "split on ASCII whitespace" such that that loop directly appends to the token set. */
+
+	/* 1. Let position be a position variable for input, initially pointing at the start of input.
+	 *    => That's the position pointer. */
+	/* 2. Let tokens be a list of strings, initially empty.
+	 *    => That's the token set. */
+
+	/* 3. Skip ASCII whitespace within input given position. */
+	position += strspn(position, ascii_whitespace);
+
+	/* 4. While position is not past the end of input: */
+	while (*position != '\0') {
+		/* 4.1. Let token be the result of collecting a sequence of code points that are not ASCII whitespace from input */
+		const char *start = position;
+		position += strcspn(position, ascii_whitespace);
+		size_t length = position - start;
+
+		/* 4.2. Append token to tokens. */
+		zend_string *token = zend_string_init(start, length, false);
+		dom_add_token(token_set, token);
+		zend_string_release_ex(token, false);
+
+		/* 4.3. Skip ASCII whitespace within input given position. */
+		position += strspn(position, ascii_whitespace);
+	}
+
+	/* 5. Return tokens.
+	 *    => That's the token set. */
+}
+
+/* https://dom.spec.whatwg.org/#concept-ordered-set-serializer */
+static char *dom_ordered_set_serializer(HashTable *token_set)
+{
+	size_t length = 0;
+	zend_string *token;
+	ZEND_HASH_MAP_FOREACH_STR_KEY(token_set, token) {
+		size_t needed_size = ZSTR_LEN(token) + 1; /* +1 for the space (or \0 at the end) */
+		if (UNEXPECTED(ZSTR_MAX_LEN - length < needed_size)) {
+			/* Shouldn't really be able to happen in practice. */
+			zend_throw_error(NULL, "Token set too large");
+			return NULL;
+		}
+		length += needed_size;
+	} ZEND_HASH_FOREACH_END();
+
+	if (length == 0) {
+		char *ret = emalloc(1);
+		*ret = '\0';
+		return ret;
+	}
+
+	char *ret = emalloc(length);
+	char *ptr = ret;
+	ZEND_HASH_MAP_FOREACH_STR_KEY(token_set, token) {
+		memcpy(ptr, ZSTR_VAL(token), ZSTR_LEN(token));
+		ptr += ZSTR_LEN(token);
+		*ptr++ = ' ';
+	} ZEND_HASH_FOREACH_END();
+	ptr[-1] = '\0'; /* replace last space with \0 */
```

Comment:
Yes, `ptr[length]` wouldn't work here because I'd have to use the pointer that points to the start, so this was easier.
Note that in C `a[b]` is syntactic sugar for `*(a+b)`, so that means that `a[b]` and `b[a]` are the same thing. In other words, I could've also written `-1[ptr]` and it would've been equivalent, albeit a bit more cursed.

Reply:
That would be pretty cursed :D